### PR TITLE
 [MB1089] [Cluster] Hazelcast issue in server startup

### PR DESCRIPTION
**Pull request by a5anka**
**Retry node ID read when checking for duplicate node IDs**

This is done to avoid getting a Null pointer exception at startup
when multiple nodes are started paralley in the cluster.

### CODE REVIEW

**Comment by sajinidesilva** [comment]
Don't we have a public jira for this?

**Comment by a5anka** [comment]
Public JIRA is https://wso2.org/jira/browse/MB-1089

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java [hunk 1]
```diff
@@ -195,6 +192,47 @@ public void start(ClusterManager manager) throws AndesException{
...
+
+        for (Member member : members) {
+            int nodeIdReadAttempts = 1;
+            String nodeIdOfMember = getIdOfNode(member);
+
+            /*
+             Node ID can be null if the node has not initialized yet. Therefore try to read the node id
+             MAX_NODE_ID_READ_ATTEMPTS times before failing.
+              */
+            while (nodeIdOfMember == null && nodeIdReadAttempts <= MAX_NODE_ID_READ_ATTEMPTS) {
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java]: Shall we edit this like  ((nodeIdOfMember == null) && (nodeIdReadAttempts <= MAX_NODE_ID_READ_ATTEMPTS)) for readability purposes?

Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java]: :+1:

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java [hunk 2]
```diff
@@ -195,6 +192,47 @@ public void start(ClusterManager manager) throws AndesException{
...
+
+                nodeIdReadAttempts++;
+                nodeIdOfMember = getIdOfNode(member);
+            }
+
+            if (nodeIdOfMember == null) {
+                throw new AndesException("Failed to read Node id of hazelcast member " + member);
+            }
+
+            if (localMember != member && nodeIdOfMember.equals(getLocalNodeIdentifier())) {
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java]: Shall we add parenthesisf or each statement inside the if condition?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java [hunk 3]
```diff
@@ -195,6 +192,47 @@ public void start(ClusterManager manager) throws AndesException{
...
+             Node ID can be null if the node has not initialized yet. Therefore try to read the node id
+             MAX_NODE_ID_READ_ATTEMPTS times before failing.
+              */
+            while (nodeIdOfMember == null && nodeIdReadAttempts <= MAX_NODE_ID_READ_ATTEMPTS) {
+
+                // Exponentially increase waiting time,
+                long sleepTime = Math.round(Math.pow(2, nodeIdReadAttempts));
+                log.warn("Node id was null for member " + member + ". Node id will be read again after "
+                         + sleepTime + " seconds.");
+                try { TimeUnit.SECONDS.sleep(sleepTime); } catch (InterruptedException ignore) {}
```
Comment (sajinidesilva) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java]: Formatting is kind of confusing.

Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java]: What do you suggest?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java (hunk 2 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/cluster/HazelcastClusterAgent.java]: :+1:

**Pull request merged**